0/Issuance.js

```diff
@@ -0,0 +1,146 @@
+const crypto = require('crypto');
+const util = require('../util');
+const factomCryptoValidation = require('factom-identity-lib/src/validation');
+const fctIdentityCrypto = require('factom-identity-lib/src/crypto');
+const fctIdentityUtil = require('factom-identity-lib/src/validation');
+const Transaction = require('./Transaction').Transaction;
+
+
+class IssuanceBuilder {
+
+    constructor(rootChainId, tokenId, sk1) {
+        if (!fctIdentityUtil.isValidIdentityChainId(rootChainId)) throw new Error("You must include a valid issuer identity Root Chain Id to issue a FAT token");
+        this._rootChainId = rootChainId;
+
+        if (tokenId === undefined || typeof  tokenId !== 'string') throw new Error('Token is a required string');
+        this._tokenId = tokenId;
+
+        if (!factomCryptoValidation.isValidSk1(sk1)) throw new Error("Supplied key is not a valid sk1 private key");
+        this._sk1 = sk1;
+
+        this._type = 'FAT-0'
+    }
+
+    name(name) {
+        if (!name) throw new Error("Token name must be defined");
+        this._name = name;
+        return this;
+    };
+
+    symbol(symbol) {
+        if (!symbol) throw new Error("Token symbol must be defined");
+        if (!new RegExp('[A-Z ]+').test(symbol)) throw new Error("Token symbol must only contain capital letters A-Z");
+        if (symbol.length == 0 || symbol.length > 4) throw new Error("Token symbol must be 1 - 4 characters");
+        this._symbol = symbol;
+        return this;
+    };
+
+    supply(supply) {
+        if (isNaN(supply)) throw new Error("Supply must be a number");
+        if (supply <= 0) throw new Error("Supply must be > 0");
+        this._supply = supply;
+        return this;
+    };
+
+    salt(salt) {
+        if (typeof salt !== 'string' || salt.length < 1) throw new Error('salt must be a non empty string, if defined (optional)');
+        this._salt = salt;
+        return this;
+    }
+
+    coinbase(transaction) {
+        if (!transaction instanceof Transaction) throw new Error("Transaction must be of type Transaction");
+        if (!this._transactions) this._transactions = [transaction];
+        else this._transactions.push(transaction);
+        return this;
+    };
+
+    build() {
+        //validate required fields
+
+        return new Issuance(this);
+    }
+}
+
+class Issuance {
+    constructor(builder) {
+
+        if (builder instanceof IssuanceBuilder) {
+            this._type = builder._type;
+            this._name = builder._name;
+            this._symbol = builder._symbol;
+            this._supply = builder._supply;
+            this._salt = builder._salt || crypto.randomBytes(32).toString('hex');
+
+            this._content = JSON.stringify(this);
```

Comment:
The `content` is the JSON content of the entry on Factom (so excluding any of the other fields declared after line 75, or others returned by the API). Note this can only be certified for issuances/txs built/constructed locally in the library because guessing of json field ordering of an external API changes the data represented in the signatures